bundles/alpha.codegen/src/alpha/codegen/isl/MemoryUtils.xtend

```diff
@@ -0,0 +1,127 @@
+package alpha.codegen.isl
+
+import fr.irisa.cairn.jnimap.barvinok.BarvinokBindings
+import fr.irisa.cairn.jnimap.isl.ISLConstraint
+import fr.irisa.cairn.jnimap.isl.ISLDimType
+import fr.irisa.cairn.jnimap.isl.ISLPWQPolynomial
+import fr.irisa.cairn.jnimap.isl.ISLSet
+
+import static extension alpha.model.util.CommonExtensions.toArrayList
+
+/** Helper methods intended for memory allocation and accessing using isl objects. */
+class MemoryUtils {
+	/**
+	 * Returns a piecewise quasi-affine polynomial indicating
+	 * the number points in the given set (i.e., its cardinality).
+	 * 
+	 * @param domain The domain to get the cardinality of. This set is not destroyed ("isl_keep").
+	 * @returns A piecewise quasi-affine polynomial that represents the cardinality of the given domain.
+	 */
+	def static ISLPWQPolynomial card(ISLSet domain) {
+		BarvinokBindings.card(domain.copy)
+	}
+	
+	/**
+	 * Returns a piecewise quasi-affine polynomial which can compute the "rank" of a point.
+	 * The "rank" is defined as the number of points which are lexicographically
+	 * less than that point.
+	 * We will compute this by building the set of such points, then using the Barvinok
+	 * library to compute the cardinality of such a set.
+	 * This will return a piecewise quasi-affine polynomial.
+	 * 
+	 * Note: the polynomial returned may look like it's undefined for the lex min of the given set.
+	 * However, this is not actually the case.
+	 * These polynomials are defined over the universe, and will evaluate to zero
+	 * if the point you're asking for lies outside the domains of each piece.
+	 * In other words: there is an implicit "default" case that always evaluates to zero
+	 * if the point being evaluated lies outside all the domains of all the pieces of the polynomial.
+	 * 
+	 * For example, the set "[N] -> {[i]: 0 <= i <= N}" (which forms a line from 0 to N)
+	 * would return the polynomial "[N,i] -> { i : 0 < i <= N }".
+	 * Per the previous paragraph, this is still valid for i=0,
+	 * as it falls under the implicit extra case that always evaluates to zero.
+	 * 
+	 * The intended use case of this is for determining the index of a point 
+	 * within a linearized array of our set. This produces a 0-based indexing.
+	 * Per the second paragraph, it may look like this polynomial is not defined for index 0,
+	 * but it actually is due to the implicit default case always evaluating to 0
+	 * if the point you're evaluating at is not in any of the domains of the pieces.
+	 * 
+	 * @param domain The domain to get the ranking polynomial for. This set is not destroyed ("isl_keep").
+	 * @returns A piecewise quasi-affine polynomial that can be used to compute the rank of a point within the given domain.
+	 */
+	def static ISLPWQPolynomial rank(ISLSet domain) {
```

Comment:
This function is worth looking at in detail, as this is what handles indexing the linearized memory.

Reply:
Just summarizing for my own understanding/confirm. For the given input set, we introduce one new index, i’, for each existing index i. Then move all preexisting indices to parameters. Then add constraints s.t. all i’ < i. Given `[N] -> { [i] : N >= 11 and 0 <= i < N }`:  
* `extendedSet` —> `[N, i] -> { [i0] : N >= 11 and 0 <= i < N }`
* `lessThan` —> `[N, i] -> { [i0] : N >= 11 and 0 <= i < N and i0 < i }`
* returned cardinality —> `[N, i] -> { i : N >= 11 and 0 < i < N }`

Reply:
Yes, that's correct. One additional point (more for future reference than anything), if you have multiple dimensions, e.g., `i` and `j`, we'll construct two sets of constrains and union those sets together: `i' < i` is one, `i' = i and j' < j` will be the other.